lib/new-admin/graphql/modules/authentication.js

```diff
@@ -0,0 +1,220 @@
+const otplib = require('otplib')
+const bcrypt = require('bcrypt')
+
+const loginHelper = require('../../login')
+const T = require('../../../time')
+const users = require('../../../users')
+const sessionManager = require('../../../session-manager')
+
+const REMEMBER_ME_AGE = 90 * T.day
+
+async function authenticateUser (username, password) {
+  const hashedPassword = await loginHelper.checkUser(username)
+  if (!hashedPassword) return null
+
+  const isMatch = await bcrypt.compare(password, hashedPassword)
+  if (!isMatch) return null
+
+  const user = await loginHelper.validateUser(username, hashedPassword)
+  if (!user) return null
+  return user
+}
```

Comment:
you are making use of the async/await format, while there's nothing intrinsically wrong with it, I think this code here is "blocking"

Reply:
As I told you earlier, I don't believe this is an issue, as it is only waiting for the result of the previous call to be computed, as the result will be needed for the call coming right after. The alternative would be a cascade of thenables, which is painful to read, with the same behavior in execution

Reply:
Discussed this with Rafael as well and changed this to promises due to how error handling on the server is working